llvm/test/Feature/Repo/repo_hash_contributions_dependencies.ll

```diff
@@ -0,0 +1,59 @@
+; This test is used to check the hash calculation for the contributions.
+;
+; Although 'contributions' and 'dependencies' respective arrows point are
+; in different direction, they can affect each other and form loops.
+;
+; The test invokes five steps:
+;  1) Compile the initial code `Inputs/repo_contributions_dependencies_loop.ll`;
+;  2) Compiled `repo_hash_contributions_dependencies.ll`.
+;  3) Dump the digest of `Var2` in %t.
+;  4) Dump the digest of `Var2` in %t1.
+;  5) Check the digests of `Var2` are different since the function `Func1` is changed.
+;
+; RUN: rm -rf %t.db
+; RUN: env REPOFILE=%t.db clang -c -O3 --target=x86_64-pc-linux-gnu-repo -x ir %S/Inputs/repo_contributions_dependencies_loop.ll -o %t
+; RUN: env REPOFILE=%t.db clang -c -O3 --target=x86_64-pc-linux-gnu-repo -x ir %s -o %t1
+; RUN: env REPOFILE=%t.db %python %S/fragments.py %t Var2 -r %t.db -d > %t.d
+; RUN: env REPOFILE=%t.db %python %S/fragments.py %t1 Var2 -r %t.db -d > %t1.d
+; RUN: not diff %t.d %t1.d
+
+target triple = "x86_64-pc-linux-gnu-elf"
+
+%"T1" = type { i8* }
+%"T2" = type { i32 (...)**, i32, i8* }
+
+@Var1 = internal constant { [1 x i8*] } { [1 x i8*] [i8* bitcast (i8* (%T1*, i8*)* @Func1 to i8*)] }, align 8
```

Comment:
This test code seems rather complex. The test itself only seems to examine `Var2`. Can any of the other code here be removed?

Reply:
I have double checked the test; I cannot remove any part of the code. This is a complex case, which current hash algorithm (in the master branch) cannot handle correctly. 
If we dumped the Contribution and Dependencies for all GOs:
```
GO Name:Func2
GOInfo:
        Initial Digest: e46ca17e0fb3526bf56c047bdfc6f544
        Contributions: [ ]
        Dependencies: [ Func1,Var1]

GO Name:llvm.global_ctors
GOInfo:
        Initial Digest: f88095266ad49d5eeff99f4577a1fa15
        Contributions: [ ]
        Dependencies: [ _GLOBAL__sub_I_test.cpp]

GO Name:Var1
GOInfo:
        Initial Digest: 478c840fb812e499fa10477bd6f86e2a
        Contributions: [ ]
        Dependencies: [ Func1]

GO Name:Func1
GOInfo:
        Initial Digest: bfba76f120769a6a88be36dd5ecc6177
        Contributions: [ ]
        Dependencies: [ ]

GO Name:__cxx_global_var_init
GOInfo:
        Initial Digest: 1c7e6256cf1a747b1037516744c1cc89
        Contributions: [ ]
        Dependencies: [ Var2,Func2]

GO Name:_GLOBAL__sub_I_test.cpp
GOInfo:
        Initial Digest: d82ff111b1fba61e916e09e1570fe280
        Contributions: [ ]
        Dependencies: [ __cxx_global_var_init]

GO Name:Var2
GOInfo:
        Initial Digest: 55546ce12e6136dc08bdaad6844334c8
        Contributions: [ __cxx_global_var_init]
        Dependencies: [ ]
```

The expected fragment of `Var2` has an xfixup to `Var1`.
```
$ pstore-dump --fragment=e9d2007ba07e5c64fcf070ef40ca30d5 clang.db
---

- file      :
      path : clang.db
      size : 4194304
  fragments :
      - digest   : e9d2007ba07e5c64fcf070ef40ca30d5
        fragment :
            - type     : data
              contents :
                  align   : 8
                  data    : !!binary |
                      AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
                  ifixups : [ ]
                  xfixups :
                      - name   : Var1
                        type   : 1
                        offset : 0
                        addend : 16
...
```
I want to test the `Var2` digest value will be changed if we modified the function `Func1`. Please notes: the current issue is the `Var2` digest keeps the same when I modified the  `Func1`.

When I modified `Func1`, the `Var1` could be changed since `Func1` is a dependency of `Var1`. If `Var1` is changed, the `Var2` could be changed by calling `Func2`.
Therefore, all these functions need to be there to produce the expecting case which shows the issue.

I tried to remove any function to get a simple test case, the simple test can be handled correctly with correct master branch hash algorithm.

Reply:
> I want to test the Var2 digest value will be changed if we modified the function Func1.

Ok, understood. What role to the static ctors play in this (__cxx_global_var_init, _GLOBAL__sub_I_test.cpp, and so on)?

Reply:
>
> When I modified Func1, the Var1 could be changed since Func1 is a dependency of Var1. If Var1 is changed, the Var2 could be changed by calling Func2.
>Therefore, all these functions need to be there to produce the expecting case which shows the issue.
>

In summary, the `Var1`, `Var2`, `func1` and `func2` are needed.

This issue is `undefined symbol`.  To show the issue, the test has to include a loop dependency between a global variable `Var2` and a global function `__cxx_global_var_init`. Therefore,  **`__cxx_global_var_init`** is needed.

Please remember, the initial issue is a linking time error: undefined symbol to `Var1`. The issue is caused by incorrect hash calculation of `Var2`. In the master branch, the `Var1` hash value is changed but `Var2` doesn't. The issue is caused by this. I need to make sure that the fragment `Var2` has a xfixup to `Var1`. 

The LLVM global variable optimisation pass makes `Var1` is referenced by `Var2` when optimising the `llvm.global_ctors` . Again, if you remove the `llvm.global_ctors`, the `Var2` has no xfixups and current master branch can handle the case. Therefore,  **_`llvm.global_ctors` and `_GLOBAL__sub_I_test.cpp`_** are needed.

Reply:
> This issue is undefined symbol. 

Yes, that was the original problem that motivated this work. However…

> The issue is caused by incorrect hash calculation of Var2.

Yes, and the test ensures that the hash changes when `Func1` is changed (a change to the align attributes from 8 to 16). This is good.

However, all of that said, the test doesn‘t attempt to provoke an undefined symbol error. The precise relationship between Func1 and Var2 is extremely difficult to determine from the test (I’ve tried!) My call for simplicity derives from my wish to understand exactly what is being tested and my concern in working out what is broken if this test fails in future.

> I need to make sure that the fragment Var2 has a xfixup to Var1.

During the hash calculation (the component being changed and the one being tested) we’re long before any external fixups exist in the compiler: that doesn’t happen until the final RepoWriter stage when relocations are being recorded. The test doesn't check the external fixups on Var2: this seems fair because we’re not changing that aspect of the compiler’s behaviour.

> The LLVM global variable optimisation pass makes `Var1` is referenced by `Var2` when optimising the `llvm.global_ctors`.

Again, this happens later in the compiler. Is this important for the hash of Var2?

Reply:
I will try to answer the following two questions you mentioned in previous comments.
>
> Again, this happens later in the compiler. Is this important for the hash of Var2?
>
Yes, it is very important for the hash of Var2. The `Var2` has zeroinitializer during the RepoMetadataGenerationPass, however its initial value changes after the Global Variable Optimizer pass. The hash algorithm needs to consider this.
The issue #25 gives the detailed explanation.

>
> The precise relationship between Func1 and Var2 is extremely difficult to determine from the test.
>
In the function `__cxx_global_var_init`, the `Func2` is used to set the `Var2` initial value. The `Func2` stores `Var1` value to `Var2`. Therefore, if the `Var1` value changes, the `Var2` value will be changed. `Var1` contains a function pointer of `Func1`. If the function `Func1` changes (e.g. alignment), the `Var1` will be changed, which will result in the change of `Var2`. Hope, this make sense to you.

Reply:
> The Var2 has zeroinitializer during the RepoMetadataGenerationPass, however its initial value changes after the Global Variable Optimizer pass. The hash algorithm needs to consider this.

Yes, that’s true. However hashing of changes made by the optimizer is not what is being fixed here: we’re interested only in the handling of dependencies and contributions. This test should try to be a specific to this aspect as possible.

Reply:
> 
> we’re interested only in the handling of dependencies and contributions. 

We should get the correct hash calculation if we handle the dependencies and contributions correctly. The test tried to check that the `Var2` hash value will be updated when the hash value of function `func1` is changed. Again, as explained as above, I cannot remove any functions in order to show the bug in the master branch.

Reply:
> We should get the correct hash calculation if we handle the dependencies and contributions correctly.

Yes. However, we’re not interested in the “correct” hash: that’s not what’s being changed here. This change ensures that changes to a contribution and/or a dependency of an object affect its hash. We can test for these changes without relying on the behaviour of the optimizer.

We need to ensure that the test for this change is as far as possible testing only this change and not other parts of the compiler whose behaviour is out of our control.

> I cannot remove any functions in order to show the bug in the master branch.

Again, you’re not trying to “show the bug” (the test as currently written doesn’t check for an undefined reference error anyway): you’re trying to verify the change that you’ve made.

Reply:
Thanks for discussing the test. I have updated the test following our discussion.